apstools/beamtime/bss_info.py

```diff
@@ -0,0 +1,459 @@
+                        count += 1
+                if count > 0:
+                    matches.append(
+                        dict(
+                            esaf=esaf, 
+                            proposal=proposal, 
+                            num_true=count, 
+                            num_esaf_badges=len(esaf_badges), 
+                            num_proposal_badges=len(findings),
+                        )
+                    )
+    return matches
+
+
+def getCurrentProposals(beamline):
+    proposals = []
+    for cycle in listRecentRuns():
+        for prop in api_bss.listProposals(beamlineName=beamline, runName=cycle):
+            prop = dict(prop)
+            prop["cycle"] = cycle
+            proposals.append(prop)
+    return proposals
+
+
+def getEsaf(esafId):
+    try:
+        record = api_esaf.getEsaf(int(esafId))
+    except dm.ObjectNotFound:
+        raise EsafNotFound(esafId)
+    return dict(record.data)
+
+
+def getProposal(proposalId, runName, beamlineName):
+    # avoid possible dm.DmException
+    if runName not in listAllRuns():
+        raise DmRecordNotFound(f"run '{runName}' not found")
+
+    if beamlineName not in listAllBeamlines():
+        raise DmRecordNotFound(f"beamline '{beamlineName}' not found")
+
+    try:
+        record = api_bss.getProposal(str(proposalId), runName, beamlineName)
+    except dm.ObjectNotFound:
+        raise ProposalNotFound(
+            f"id={proposalId}"
+            f" run={runName}"
+            f" beamline={beamlineName}"
+            )
+    return dict(record.data)
+
+
+def iso2datetime(isodate):
+    return datetime.datetime.fromisoformat(isodate)
+
+
+def listAllBeamlines():
+    if "beamlines" not in _cache_:
+        _cache_["beamlines"] = [
+            entry["name"]
+            for entry in api_bss.listBeamlines()
+        ]
+    return _cache_["beamlines"]
+
+
+def listAllRuns():
+    if "cycles" not in _cache_:
+        _cache_["cycles"] = sorted([
+            entry["name"]
+            for entry in api_bss.listRuns()
+        ])
+    return _cache_["cycles"]
+
+
+def listRecentRuns(quantity=6):
+    # 6 runs is the duration of a user proposal
+    tNow = datetime.datetime.now()
+    runs = [
+        run["name"]
+        for run in api_bss.listRuns()
+        if iso2datetime(run["startTime"]) <= tNow
+    ]
+    return sorted(runs, reverse=True)[:quantity]
+
+
+def printColumns(items, numColumns=5, width=10):
+    n = len(items)
+    rows = n // numColumns
+    if n % numColumns > 0:
+        rows += 1
+    for base in range(0, rows):
+        row = [
+            items[base+k*rows]
+            for k in range(numColumns) 
+            if base+k*rows < n]
+        print("".join([f"{s:{width}s}" for s in row]))
+
+
+def trim(text, length=40):
+    if len(text) > length:
+        text = text[:length-3] + "..."
+    return text
+
+
+class DmRecordNotFound(Exception): ...
+class EsafNotFound(DmRecordNotFound): ...
+class ProposalNotFound(DmRecordNotFound): ...
```

Comment:
Codacy found an issue: [Statement seems to have no effect](https://app.codacy.com/gh/BCDA-APS/apstools/pullRequest?prid=5764884)